source/gridfs/gridfs-spec.md

```diff
@@ -0,0 +1,1021 @@
+
+- **bucketName:** the name of this GridFS bucket. The files and chunks collection for this GridFS bucket are prefixed by
+  this name followed by a dot. Defaults to "fs". This allows multiple GridFS buckets, each with a unique name, to exist
+  within the same database.
+- **chunkSizeBytes:** the number of bytes stored in chunks for new user files added through this GridFSBucket object.
+  This will not reformat existing files in the system that use a different chunk size. Defaults to 255 KiB.
+
+IF a driver supports configuring readConcern, readPreference or writeConcern at the database or collection level, then
+GridFSBucket objects MUST also allow the following options to be configurable:
+
+- **readConcern:** defaults to the read concern on the parent database (or client object if the parent database has no
+  read concern).
+- **readPreference:** defaults to the read preference on the parent database (or client object if the parent database
+  has no read preference).
+- **writeConcern:** defaults to the write concern on the parent database (or client object if the parent database has no
+  write concern).
+
+The following option is transitional:
+
+- **disableMD5:** this allows users to disable MD5 when operating under FIPS restrictions. It is provided to allow a
+  transition period as drivers remove MD5 support. Until a driver removes MD5 support, drivers MUST support this option.
+  Following a driver's normal feature removal cycle, when MD5 support is removed, this option MUST be removed or
+  otherwise made into a no-op option.
+
+GridFSBucket instances are immutable. Their properties MUST NOT be changed after the instance has been created. If your
+driver provides a fluent way to provide new values for properties, these fluent methods MUST return new instances of
+GridFSBucket.
+
+### Indexes
+
+For efficient execution of various GridFS operations the following indexes MUST exist:
+
+- an index on { filename : 1, uploadDate : 1 } on the files collection
+- a unique index on { files_id : 1, n : 1 } on the chunks collection
+
+Normally we leave it up to the user to create whatever indexes they see fit, but because GridFS is likely to be looked
+at as a black box we should create these indexes automatically in a way that involves the least amount of overhead
+possible.
+
+#### Before read operations
+
+For read operations, drivers MUST assume that the proper indexes exist.
+
+#### Before write operations
+
+Immediately before the **first** write operation on an instance of a GridFSBucket class is attempted (and not earlier),
+drivers MUST:
+
+- determine if the files collection is empty using the primary read preference mode.
+- and if so, create the indexes described above if they do not already exist
+
+To determine whether the files collection is empty drivers SHOULD execute the equivalent of the following shell command:
+
+> > db.fs.files.findOne({}, { \_id : 1 })
```

Comment:
Use three backticks to code-fence this line instead of the leading `>` here. The inner `>` should be preserved since it represents the shell prompt.

Reply:
Done